docs/secrets.md

```diff
@@ -0,0 +1,138 @@
+sha256 - sha256 hashes your secret; to be used as a secondary function i.e. `|randomstr|sha256`. You can optionally pass a salt i.e `|randomstr|sha256:salt` -> becomes `sha256("salt:<generated random string>")`
+reveal - Decrypts a secret; to be used as a secondary function, useful for reuse of a secret like for different encodings i.e `|reveal:path/to/secret|base64`
+rsa - Generates an RSA 4096 private key (PKCS#8). You can optionally pass the key size i.e. `|rsa:2048`
+rsapublic - Derives an RSA public key from a revealed private key i.e. `|reveal:path/to/encrypted_private_key|rsapublic`
+```
+
+*Note*: If you use `|reveal:/path/secret`, when changing the `/path/secret` file make sure you also delete any secrets referencing `/path/secret` so kapitan can regenerate them.
+
+#### 3. Reference your secrets in your classes/targets and run `kapitan compile`
+
+Secrets can be referenced in the format `?{<secret_type>:path/to/secret/file}`.
+
+For example, assume for now that your GPG-encrypted secret is already stored in a file at `targets/secrets/mysql_password`. This can be referenced in the inventory in the following format:
+
+```yaml
+users:
+  root:
+    # If 'secrets/targets/${target_name}/mysql/password' doesn't exist, we can automatically generate a random b64-encoded password as follows
+    password: ?{gpg:targets/${target_name}/mysql/password|randomstr|base64}
+```
+
+During compile, kapitan will search for the path `targets/${target_name}/mysql/password`. Should it not exist, then it will automatically generate a random base64 password and save it to that path.
+
+#### 4. Reveal and use the secrets
+
+You can reveal the secrets referenced in the outputs of `kapitan compile` via:
+
+```
+$ kapitan secrets --reveal -f path/to/rendered/template
+```
+
+For example, `compiled/minikube-mysql/manifests/mysql_secret.yml` with the following content:
+
+```yaml
+apiVersion: v1
+data:
+  MYSQL_ROOT_PASSWORD: ?{gpg:targets/minikube-mysql/mysql/password:ec3d54de}
+  MYSQL_ROOT_PASSWORD_SHA256: ?{gpg:targets/minikube-mysql/mysql/password_sha256:122d2732}
+kind: Secret
+metadata:
+  annotations: {}
+  labels:
+    name: example-mysql
+  name: example-mysql
+  namespace: minikube-mysql
+type: Opaque
+```
+
+can be revealed as follows:
+
+```
+$ kapitan secrets --reveal -f compiled/minikube-mysql/manifests/mysql_secret.yml
+```
+
+This will substitute the referenced secrets with the actual decrypted secrets stored at the referenced paths and display the file content.
+
+## Secret sub-variables
+
+As illustrated above, one file corresponds to one secret. It is now possible for users who would like to reduce the decryption overhead to manually create a yaml file that contains multiple secrets, each of which can be referenced by its object key. For example, consider the secret file `secrets/mysql_secrets`:
+
+```yaml
+mysql_passwords:
+	secret_foo: hello_world
```

Comment:
These are marked as tabs in github. This might be a good chance to change these to just spaces as we have everywhere else. :)